File: examples/host/c/01_hello_d1/main.c
/*
 * D1 C 语言示例 —— 01_hello_d1
 *
 * 本示例演示 D1 C API 的完整用法：
 *   1. 获取版本号
 *   2. 初始化 D1 运行时
 *   3. 设置消息处理器（演示 API：Publish、Call、CacheSet/CacheGet、DBQuery）
 *   4. 启动 D1
 *   5. 阻塞等待退出（Ctrl+C）
 *
 * 编译时需要链接 libd1.so，详见 CMakeLists.txt。
 * D1 动态库依赖: >= v1.5.0
 */

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../../../lang/c/d1.h"

/* ─────────────────────────────────────────────────────────
 * 消息处理器 — 演示所有 D1 API 的调用方式
 * ───────────────────────────────────────────────────────── */

static int on_request(uint64_t task_id,
                      const char* method,
                      const char* params,
                      int params_len,
                      char** out_result,
                      int* out_len,
                      char** out_error) {
    printf("[处理器] task_id=%lu, method=%s\n", (unsigned long)task_id, method);

    /* ── 1. Publish — 发送单向消息（无回复） ── */
    const char* pub_data = "{\"temp\":25.5,\"unit\":\"celsius\"}";
    int ret = D1_Publish(task_id, "mqtt_client", "sensor.data",
                         pub_data, (int)strlen(pub_data));
    printf("  D1_Publish → %d\n", ret);

    /* ── 2. CacheSet — 写入缓存 ── */
    const char* cache_val = "{\"name\":\"Alice\",\"role\":\"admin\"}";
    ret = D1_CacheSet(task_id, "user:42", cache_val, (int)strlen(cache_val), 3600);
    printf("  D1_CacheSet → %d\n", ret);

    /* ── 3. CacheGet — 读取缓存 ── */
    char* cached = NULL;
    int cached_len = 0;
    ret = D1_CacheGet(task_id, "user:42", &cached, &cached_len);
    if (ret == 0 && cached) {
        printf("  D1_CacheGet → %.*s\n", cached_len, cached);
        D1_Free(cached);
    }

    /* ── 4. Call — 同步调用（阻塞等待结果） ── */
    char* call_result = NULL;
    int call_len = 0;
    char* call_error = NULL;
    ret = D1_Call(task_id, "script", "api_handler", "get_user",
                  "{\"id\":123}", 10, 5,
                  &call_result, &call_len, &call_error);
    if (ret == 0 && call_result) {
        printf("  D1_Call → %.*s\n", call_len, call_result);
        D1_Free(call_result);
    } else if (call_error) {
        printf("  D1_Call error → %s\n", call_error);
        D1_Free(call_error);
    }

    /* ── 5. DBQuery — 数据库查询 ── */
    char* db_result = NULL;
    int db_len = 0;
    ret = D1_DBQuery(task_id, "SELECT * FROM users LIMIT 1", 30,
                     &db_result, &db_len);
    if (ret == 0 && db_result) {
        printf("  D1_DBQuery → %.*s\n", db_len, db_result);
        D1_Free(db_result);
    }

    /* ── 6. 返回响应 ── */
    const char* reply = "{\"status\":\"ok\",\"msg\":\"hello from C handler\"}";
    *out_result = (char*)malloc((size_t)strlen(reply));
    memcpy(*out_result, reply, strlen(reply));
    *out_len = (int)strlen(reply);
    *out_error = NULL;
    return 0;
}

/* ─────────────────────────────────────────────────────────
 * 主函数
 * ───────────────────────────────────────────────────────── */

int main(int argc, char* argv[]) {
    (void)argc; (void)argv;

    printf("===== D1 C Hello World =====\n");

    /* 1. 获取版本号 */
    const char* version = D1_Version();
    printf("D1 Version: %s\n", version ? version : "(null)");

    /* 2. 初始化 D1 运行时 */
    if (D1_Init(NULL) != 0) {
        fprintf(stderr, "D1_Init failed\n");
        return EXIT_FAILURE;
    }
    printf("D1_Init OK\n");

    /* 3. 设置默认消息处理器 */
    D1_SetOnRequest(on_request);
    printf("Default handler registered\n");

    /* 4. 启动 D1 运行时 */
    if (D1_Start() != 0) {
        fprintf(stderr, "D1_Start failed\n");
        return EXIT_FAILURE;
    }
    printf("D1 running (press Ctrl+C to exit)\n");

    /* 5. 阻塞等待退出 */
    D1_WaitStop();
    printf("D1 stopped, exiting.\n");
    return EXIT_SUCCESS;
}

File: examples/host/python/01_hello_d1/main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
示例: D1 Hello World (Python)

本示例演示 D1 Python SDK 的完整用法:
  1. 获取版本号
  2. 初始化 D1 运行时
  3. 设置默认消息处理器（演示 API：Publish、Call、CacheSet/CacheGet、DBQuery）
  4. 启动 D1
  5. 阻塞等待退出（Ctrl+C）

使用前请确保已将 libd1.so 和 d1.h 放入 deps/ 目录。

运行方式:
    cd examples/host/python/01_hello_d1
    python3 main.py
"""

import json
import os
import sys

# 将 lang/ 目录加入 Python 路径
_sdk_path = os.path.normpath(
    os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..", "lang", "python")
)
sys.path.insert(0, _sdk_path)

from d1 import D1


def default_handler(task_id, method, params):
    """默认消息处理器。

    演示所有 D1 API 的调用方式：
    - Publish: 发送单向消息
    - CacheSet/CacheGet: 缓存读写
    - Call: 同步调用
    - DBQuery: 数据库查询
    """
    params_str = params.decode("utf-8") if params else ""
    print(f"[Handler] taskID={task_id} | method={method} | params={params_str}")

    # 1. Publish — 发送单向消息（无回复）
    pub_data = json.dumps({"temp": 25.5, "unit": "celsius"})
    d1.publish(task_id, "mqtt_client", "sensor.data", pub_data)
    print("  D1_Publish -> OK")

    # 2. CacheSet — 写入缓存
    cache_val = json.dumps({"name": "Alice", "role": "admin"})
    d1.cache_set(task_id, "user:42", cache_val, 3600)
    print("  D1_CacheSet -> OK")

    # 3. CacheGet — 读取缓存
    cached = d1.cache_get(task_id, "user:42")
    print(f"  D1_CacheGet -> {cached.decode() if cached else 'null'}")

    # 4. Call — 同步调用（阻塞等待结果）
    # kind: 0=conn, 1=default, 2=script, 3=service, 4=exec
    result = d1.call(task_id, 2, "api_handler", "get_user",
                     json.dumps({"id": 123}), 5)
    if result["error"] is None:
        print(f"  D1_Call -> {result['payload'].decode()}")
    else:
        print(f"  D1_Call error -> {result['error']}")

    # 5. DBQuery — 数据库查询
    rows = d1.db_query(task_id, "SELECT * FROM users LIMIT 1")
    print(f"  D1_DBQuery -> {rows.decode() if rows else 'null'}")

    # 返回响应
    return json.dumps({"status": "ok", "msg": "hello from Python handler"}).encode()


def main():
    """主函数: D1 Hello World 示例。"""
    print("===== D1 Python Hello World =====")

    d1 = D1()

    # 1. 输出版本信息
    ver = d1.version()
    print(f"D1 Version: {ver}")

    # 2. 初始化 D1 运行时
    config_path = os.environ.get("D1_CONFIG", None)
    d1.init(config_path)
    print("D1::init OK")

    # 3. 设置默认消息处理器
    d1.set_on_request(default_handler)
    print("Default handler registered")

    # 4. 启动 D1 运行时
    d1.start()
    print("D1::start OK, running (press Ctrl+C to exit)")

    # 5. 阻塞等待退出
    d1.wait_stop()
    print("D1 stopped, exiting.")


if __name__ == "__main__":
    main()